YAML format documentation

speth opened the pull request:
* Move all CTI-specific documentation out of the Science section and into the CTI guide. The Science docs now link to the CTI docs where appropriate.
* Add an analogue to the CTI "Defining Phases" guide for the YAML format
* Link from relevant sections of the "Science" docs to the YAML API docs

I have also pushed this PR branch to the `testing` branch, so this is currently live at https://testing.cantera.org (specifically, https://testing.cantera.org/tutorials/yaml/defining-phases.html)

speth commented on pages/tutorials/yaml/phases.rst:
```diff
@@ -0,0 +1,294 @@
+    - `compound-lattice <https://cantera.org/documentation/dev/sphinx/html/yaml/phases.html#compound-lattice>`__
+    - `constant-density <https://cantera.org/documentation/dev/sphinx/html/yaml/phases.html#constant-density>`__
+    - `Debye-Huckel <https://cantera.org/documentation/dev/sphinx/html/yaml/phases.html#debye-huckel>`__
+    - `edge <https://cantera.org/documentation/dev/sphinx/html/yaml/phases.html#edge>`__
```
Do we have a better way of linking into the dev docs? Using ``` :ref:`section-name` ``` only links to the stable docs. Or even, how does one add a new magic command like ``` :yaml:`section-name` ```?

bryanwweber replied:
To add a new magic command like `:yaml:`, you need to edit the `parse_docs` plugin: https://github.com/Cantera/cantera-website/blob/master/plugins/parse_docs.py You can more-or-less copy the implementation there to add a new identifier, in other words, add a `yaml_targets` dictionary to the `site` instance, and add the `yaml` folder to the loop at the bottom. That will get the targets processed into the cache. I think it only works for the current version docs, you might need to add a special case for YAML since its in the dev docs that we'll have to remove when we release the new version.

The section linking is done in the `ref.py` and `process_ref_targets.py` plugins. For the most part, the `ref.py` should just work if it finds a `:ref:` directive, so long as the location is stored in the cached `ref_targets` dictionary. So if you need to add new files/locations in the Science docs (the API docs are handled by the `parse_docs.py` plugin), just make sure that folder is processed in `process_ref_targets.py`.

I really need to document this...

speth replied:
Well, I managed to get `parse_docs.py` to read through the dev docs, and I added an entry in `rest_class.py` so that the new role `ctidev:class` doesn't generate errors when used, but for some reason it produces links to the 2.4 docs instead of the dev. docs.

speth replied:
Looking at `process_ref_targets.py`, I don't understand how it decides what folders to process. Does it process `:ref:` targets in the (stable) API docs? And how would this work if we wanted to have targets in the dev docs. Would this require another role to be introduced, e.g. `:devref:`?

speth commented on pages/tutorials/yaml/phases.rst:
```diff
@@ -0,0 +1,294 @@
+
+    - `gas <https://cantera.org/documentation/dev/doxygen/html/de/dae/classCantera_1_1GasKinetics.html#details>`__
+    - `surface <https://cantera.org/documentation/dev/doxygen/html/d1/d72/classCantera_1_1InterfaceKinetics.html#details>`__
+    - `edge <https://cantera.org/documentation/dev/doxygen/html/d0/df0/classCantera_1_1EdgeKinetics.html#details>`__
```
This seems even less likely, but is there a better way to link to Doxygen content? I guess this question would be moot if we moved the Doxygen content into Sphinx using Breathe or something similar.

bryanwweber replied:
Nope, sorry, see #16

speth commented:
I'm having a little trouble with the `:ref:`-style links to the YAML documentation for cases where there are capital letters in the section name (e.g. `pressure-dependent-Arrhenius`). It seems that to get the "website" to link to these targets, it's necessary to use all lowercase letters, but the case sensitive name within the Sphinx docs. Any idea if there's an easy fix for this, @bryanwweber?

bryanwweber commented:
I don't see any `str.lower`s anywhere 😢 (I assume you checked for those anyways 😂 ). My first guess is that this is buried somewhere deep in the docutils internals. Can you check the `cache` folder for `ref_targets` and see how the `labelid` is saved in there?

speth commented:
They've already been lowercased at that point.

speth commented:
Aha, I figured it out - this goes the other way around. According to the Docutils docs (http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#reference-names) the names are supposed to be case insensitive, and be normalized to lowercase. It shouldn't be looking for the capitalized reference name in the first place.

speth commented:
I think this is ready for review. I've also pushed this to the `testing` branch, so this now corresponds to the content available at https://testing.cantera.org.

@ischoegl I'd like to request your review here as well -- with this documentation completed, this is a good opportunity for feedback on the YAML format in general.

ischoegl commented:
@speth ... will have a look once my official teaching (and grading) duties are taken care of. 

Regarding the YAML format, I think it is overall very well executed (kudos!). I know that I have a question about chemically activated reactions (implementation is almost identical to falloff, and it may be simpler to just implement a binary flag for the latter). More later.

bryanwweber commented on pages/tutorials/yaml/species.rst:
```diff
@@ -0,0 +1,209 @@
+molecules in the solvation cage surrounding it.
+
+For surface species, it is possible for the ``composition`` mapping to be empty,
+in which case it is composed of nothing, and represents an empty surface site.
```
Could clarify what "it" is

bryanwweber commented on pages/tutorials/yaml/yaml-format.rst:
```diff
@@ -0,0 +1,374 @@
+Dimensional Values
+==================
+
+Many fields have numerical values that represent dimensional quantities -- a
```
```suggestion
Many fields have numerical values that represent dimensional quantities---a
```

bryanwweber commented on pages/tutorials/yaml/yaml-format.rst:
```diff
@@ -0,0 +1,374 @@
+another mapping applies to that mapping and any nested mappings or sequences, and overrides higher-level ``units`` mappings:
+
+.. code:: yaml
+
```
Extra lines here

bryanwweber commented on pages/tutorials/yaml/yaml-format.rst:
```diff
@@ -0,0 +1,374 @@
+
+During processing of an input file, errors may be encountered. These could be
+syntax errors, or could be ones that are flagged as errors by Cantera due to
+some apparent inconsistency in the data--an unphysical value, a species that
```
```suggestion
some apparent inconsistency in the data---an unphysical value, a species that
```

bryanwweber commented on pages/tutorials/yaml/yaml-format.rst:
```diff
@@ -0,0 +1,374 @@
+Cantera Errors
+--------------
+
+Now let's consider the other class of errors -- ones that Cantera itself
```
```suggestion
Now let's consider the other class of errors, ones that Cantera itself
```

bryanwweber commented on pages/tutorials/yaml/yaml-format.rst:
```diff
@@ -0,0 +1,374 @@
+--------------
+
+Now let's consider the other class of errors -- ones that Cantera itself
+detects. Continuing the example above, suppose that the misspelling is
```
```suggestion
detects. Continuing the example above, suppose that the missing colon is
```

bryanwweber commented on pages/tutorials/yaml/yaml-format.rst:
```diff
@@ -0,0 +1,374 @@
+importing any species that contain undeclared elements, instead of flagging them
+as errors. You can also specify that reactions containing undeclared species
+(also usually an error) should be skipped. This allows you to very easily
+extract a portion of a large reaction mechanism.
```
This is covered in another place, right? It seems kind of out of place here, like dropping into some specific detail when most of the rest of the document has been about more general syntax-type stuff.

ischoegl replied:
Is there a way to illustrate this for a simpler example for this debugging illustration? I.e. something short enough so the YAML input can be displayed?

speth replied:
Agreed, this is getting way ahead, into the details of the `phase` definition. I've opted to remove this paragraph.

ischoegl commented on pages/tutorials/yaml/species.rst:
```diff
@@ -0,0 +1,209 @@
+=======
+
+A species entry in Cantera is used to specify the name, composition,
+thermodynamic, and transport properties of an individual species.
```
Should Cantera/enhancements#14 be considered for 2.5.0?

speth replied:
Sure, I'd consider including an implementation if someone made a PR, but I don't think it's a blocker for the release.

ischoegl commented on pages/tutorials/yaml/species.rst:
```diff
@@ -0,0 +1,209 @@
+parameterization used specified by the ``model`` field of the
+``equation-of-state`` field. The models available are:
+
+    - :ref:`constant-volume <sec-yaml-eos-constant-volume>`
```
Would it make sense to clarify some of the acronyms here (e.g. HKFT is not clear). Same goes for other bullet lists of hyperlinks ... in many cases, names are not self-explanatory and it requires multiple clicks to figure out what phase / EOS definitions entail. Is there a better solution?

speth replied:
I struggled with this a bit, since I'm trying to avoid duplicating information here that belongs either in the YAML API docs or documentation that is better left for the "Science" section (never mind that this is in many places missing). I ended up adding a very short description for each item in these lists, which hopefully makes this more useful as a tutorial.

ischoegl commented on pages/science/reactions.rst:
```diff
@@ -175,7 +167,11 @@ given by:
+:cti:class:`SRI` directive, or in the YAML format using the
+:ref:`SRI <sec-yaml-falloff>` field in the entry.
 
 Chemically-Activated Reactions
```
**Note:** Looks like in some instances, the lines I'm referring to appear at the bottom! (apologies)

speth replied:
I'm going to leave this as-is for now. Obviously, depending on the resolution of that issue, this documentation may need to be revised.

speth commented on pages/tutorials/yaml/phases.rst:
```diff
@@ -0,0 +1,382 @@
+.. slug: phases
```
I don't actually understand what the slug does in Nikola. It seems like it just matches the file name, with a few rare exceptions. The `phases` slug is already duplicated between `pages/science/phases.rst` and `pages/tutorials/cti/phases.rst`.

speth replied:
But isn't it the point of the directory structure to provide that organization? These slugs aren't (AFAICT) used as globally unique identifiers in the same way as the Sphinx labels, so I don't see the benefit of uniqueness.

bryanwweber replied:
I think they are used as globally unique identifiers, which is why Nikola warns about duplicates. I believe the `doc` shortcut is able to link directly to a file given its slug. That way the structure of the source/HTML can change but the slug (and link) remain correct.

speth replied:
I don't see any warnings from Nikola about these. I did not realize the `:doc:` role linked via these slugs. Maybe you only get the warning if you try to make an ambiguous link? I guess the examples for linking that I had been following were those that linked directly to HTML pages, e.g. ``` `Installation section </install/index.html>`__```. I do think that having URLs like `/science/science-phases.html` does seem like a downgrade compared to the current `/science/phases.html`.

bryanwweber replied:
Hmmm, I thought there were warnings. Maybe some extra output has to be enabled to see them, or maybe I resolved it earlier somehow. That's a fair point about the HTML names. If there's no warnings being generated, then it's fine to leave this as-is and we can fix it later if necessary. (See here for the `doc` role: https://getnikola.com/handbook.html#doc)

speth commented on pages/tutorials/yaml/phases.rst:
```diff
@@ -0,0 +1,382 @@
+If a kinetics model has been specified, reactions may be added to the phase. By
+default, all reactions from the ``reactions`` section of the input file will be
+added. Equivalently, the ``reactions`` entry may be specified as the string
+``all``.
```
This reflected how things worked before. I have updated it to reflect the changed behavior introduced in https://github.com/Cantera/cantera/pull/773/commits/9451ac98aa04d282ca040f851636fc79377c9055.

speth commented on pages/tutorials/yaml/phases.rst:
```diff
@@ -0,0 +1,382 @@
+      thermo: ideal-surface
+      kinetics: surface
+      reactions: [graphite-anode-reactions]
+      species: [{graphite-anode-species: [(int)]}]
```
Oops, it should just be `[{anode-species: all}]`. This got messed up when I split up the different `species` sections.

speth commented on pages/science/reactions.rst:
```diff
@@ -289,25 +265,16 @@ are reduced temperatures and reduced pressures which map the ranges
+:cti:class:`chebyshev_reaction` entry, or in the YAML format using the
+:ref:`Chebyshev <sec-yaml-Chebyshev>` reaction ``type``.
+
 Surface Reactions
```
Yes, there certainly should, but I'm not trying to resolve all of the deficiencies of the "science" section of the website in this PR.

speth commented on pages/science/species.rst:
```diff
@@ -107,26 +91,21 @@ Species Transport Coefficients
-conversion. The numeric field values should all be entered as pure numbers, with
-no attached units string.
+effect of each species on the transport properties of the phase. Currently,
+ideal-gas transport property models are implemented.
```
Well, the only ones for which species-level transport coefficients are specified are for gases.

speth commented on pages/tutorials/input-files.rst:
```diff
@@ -14,24 +14,26 @@
 - Convert a pre-existing mechanism from Chemkin (CK) format to Cantera (CTI) format
 - Create your own CTI file, either from scratch (not recommended) or by editing an existing file
+- Create your own YAML file from scratch or by editing an existing file *(New in
+  Cantera 2.5)*
```
Yes, that was missing. I've added a new section with a `ck2yaml` tutorial, which is essentially a duplicate of the `ck2cti` tutorial.

speth commented on pages/tutorials/yaml/phases.rst:
```diff
@@ -0,0 +1,382 @@
+
+- ``temperature`` or ``T``
+- ``pressure`` or ``P``
+- ``enthalpy`` or ``H``
```
I'll update the discussion of `Q` pending the resolution of https://github.com/Cantera/cantera/pull/784.

The naming logic was to use the uppercase names for consistency with the state setters like `HPX`, which use capital letters, rather than the individual properties like `h`. We could make those particular names case insensitive.

speth commented on pages/tutorials/yaml/phases.rst:
```diff
@@ -0,0 +1,382 @@
+    graphite-anode-reactions:
+    - equation: LiC6 <=> Li+(e) + C6
+      rate-constant: [5.74, 0.0, 0.0]
+      beta: 0.4
```
There's a lot of unexplored territory in electrochemistry in Cantera right now. I think as these capabilities are expanded, it will be easier to see what the best way of structuring the input is. So for the time being, I'm inclined to keep this as-is. There's nothing that would prevent us from requiring an explicit `type: electrochemical` field for electrochemical reactions in the future.

speth commented on pages/tutorials/yaml/reactions.rst:
```diff
@@ -0,0 +1,241 @@
+    - :ref:`elementary <sec-yaml-elementary>`
+    - :ref:`three-body <sec-yaml-three-body>`
+    - :ref:`falloff <sec-yaml-falloff>`
+    - :ref:`chemically-activated <sec-yaml-chemically-activated>`
```
Pending https://github.com/Cantera/cantera/issues/751

speth commented:
I think I've addressed all the issues comments except for a few related to open PRs on `Cantera/cantera`. I've also pushed the updated branch to https://testing.cantera.org.

bryanwweber commented on pages/compiling/installation-reqs.rst:
```diff
@@ -84,7 +84,7 @@ Conda Requirements
   .. code:: bash
 
-     conda create --name cantera python=3 scons cython boost numpy
+     conda create --name cantera python=3 scons cython boost numpy ruamel_yaml
```
Technically, building Cantera no longer requires installing `ruamel_yaml` per Cantera/cantera#768. Certainly, using the conversion scripts requires `ruamel_yaml` (or `ruamel.yaml`).

ischoegl replied:
I'd leave `ruamel_yaml` in as anything else is more difficult to explain. Users should expect getting the ability to convert mechanisms except for a 'minimal installation' ... the only case I can currently anticipate would be a server install without python3 (where `ck2yaml` fails anyways).

speth replied:
Compiling it may not, but running the test suite does.

ischoegl replied:
Cantera/cantera#768 sidelines `ck2yaml`, so python 2 won't be a problem?

speth replied:
I was responding to @bryanwweber's comment. This bit of the instructions is targeting Python 3, anyway.

bryanwweber replied:
That's true about the test suite, ok this change makes sense then.

ischoegl commented:
One easily addressed comment: I noticed that there's no YAML input file link in the 'Other Documentation' box on the [index page](https://testing.cantera.org/documentation/index.html).

speth commented:
> One easily addressed comment: I noticed that there's no YAML input file link in the 'Other Documentation' box on the [index page](https://testing.cantera.org/documentation/index.html).
> 
> [Also, [searching for `yaml` does not yield results](https://testing.cantera.org/documentation/docs-2.4/sphinx/html/search.html?q=yaml&check_keywords=yes&area=default), although this may be due to `testing` status.]

Even on the "testing" site, that main API docs page is for the stable release. If you follow the "Development Version" link further down the page under "Need something else?", the search will find all of the YAML API docs for the development version, and you will see links to YAML documentation under the "Other Documentation" section.

ischoegl commented:
> Even on the "testing" site, that main API docs page is for the stable release. 

:+1: Ok. That was not obvious.

bryanwweber commented on pages/tutorials/yaml/yaml-format.rst:
```diff
@@ -0,0 +1,366 @@
+
+A sequence of multiple items is specified by separating the items by commas and
+enclosing them in square brackets. The individual items can have any type --
+strings, integers, floating-point numbers (or even entries or other lists).
```
```suggestion
strings, integers, floating-point numbers, mappings, or nested lists.
```

bryanwweber commented on pages/tutorials/yaml/yaml-format.rst:
```diff
@@ -0,0 +1,366 @@
+   elements: [O, H, C, N, Ar]
+   temperature-ranges: [200.0, 1000.0, 3500.0]
+
+Sequences can also be written in "block" style, using one line for each item in
```
```suggestion
The syntax using square brackets to define a list is called **flow style**. Sequences can also be written in **block style**, using one line for each item in
```

bryanwweber commented on pages/tutorials/yaml/yaml-format.rst:
```diff
@@ -0,0 +1,366 @@
+
+A mapping is a container consisting of key--value pairs. The keys in a mapping
+must be unique. Like sequences, there are two ways to write a mapping. In the
+"flow" style, the mapping is enclosed in curly brackets, colons (followed by
```
```suggestion
**flow style**, the mapping is enclosed in curly brackets, colons followed by
```

bryanwweber commented on pages/tutorials/yaml/yaml-format.rst:
```diff
@@ -0,0 +1,366 @@
+
+   composition: {H: 2, C: 1, O: 1}
+
+In the "block" style, each key is written on a new line, followed by a colon.
```
```suggestion
In the **block style**, each key is written on a new line, followed by a colon.
```

bryanwweber commented on pages/tutorials/yaml/yaml-format.rst:
```diff
@@ -0,0 +1,366 @@
+
+All keys in Cantera YAML files are treated as strings. A Cantera YAML file is
+itself a mapping, usually in the block style. We refer to the keys in this
+top-level mapping as the "sections" of the input file.
```
```suggestion
top-level mapping as the **sections** of the input file.
```

speth replied:
So do you thing `section` be emphasized where it's used elsewhere with this meaning, then, e.g. several places in `tutorials/yaml/phases.rst` and `tutorials/yaml/species.rst`?

bryanwweber commented on pages/tutorials/yaml/yaml-format.rst:
```diff
@@ -0,0 +1,366 @@
+---------------------
+
+A common structure in Cantera input files is a nested sequence of mappings. This
+can be written in the "block" style as:
```
```suggestion
can be written in the block style as:
```

speth replied:
I'm assuming this should be `**block style**` based on your other comments.

bryanwweber commented on pages/tutorials/yaml/yaml-format.rst:
```diff
@@ -0,0 +1,366 @@
+pressure, or a density, for example. If these are entered without specifying the
+units, the default units (set by the ``units`` directive) will be used. However,
+it is also possible to specify the units for each individual dimensional
+quantity (unless stated otherwise). All that is required is to write the units
```
```suggestion
quantity, unless stated otherwise. All that is required is to write the units
```

bryanwweber commented on pages/tutorials/yaml/yaml-format.rst:
```diff
@@ -0,0 +1,366 @@
+The top part of the error message shows the chain of functions that were called
+before the error was encountered. For the most part, these are internal Cantera
+functions not of direct concern here. The relevant part of this error message is
+the part between the lines of asterisks.  This message says that the YAML parser
```
```suggestion
the part between the lines of asterisks. This message says that the YAML parser
```

bryanwweber commented on pages/tutorials/yaml/yaml-format.rst:
```diff
@@ -0,0 +1,366 @@
+
+it may be imported successfully.
+
+
```
```suggestion
```

speth commented on pages/compiling/installation-reqs.rst:
```diff
@@ -509,6 +509,10 @@ Windows Requirements
 
+   * Ruamel.yaml
+
+     * `pip install ruamel.yaml`
```
For consistency, I went with a code block instead.

speth commented on pages/science/reactions.rst:
```diff
@@ -70,30 +76,9 @@ where :math:`C_{\mathrm{k}}` is the concentration of species :math:`\mathrm{k}`.
 collision efficiency can be absorbed into the rate coefficient :math:`k_f(T)`, the default collision
 efficiency is 1.0.
 
-A three-body reaction may be defined using the :cti:class:`three_body_reaction`
```
If it's not part of the diff, then that's what was here before. You're welcome to make a commit implementing this change.

bryanwweber replied:
I know, I thought it would be easy enough for you to make the change here since you have the file already open to make other changes and it is related to some of the other areas that have changed.

speth commented on pages/tutorials/yaml/phases.rst:
```diff
@@ -0,0 +1,418 @@
+  liquid electrolyte which obeys the Pitzer formulation for nonideality
+- :ref:`ideal-gas <sec-yaml-ideal-gas>`: A mixture which obeys the ideal gas law
+- :ref:`ideal-gas-VPSS <sec-yaml-ideal-gas-VPSS>`: An ideal gas; Uses "variable
+  pressure standard state" methods for calculating thermodynamic properties
```
Still waiting on a review of https://github.com/Cantera/cantera/pull/787

speth commented on pages/tutorials/yaml/phases.rst:
```diff
@@ -0,0 +1,418 @@
+- :ref:`ideal-condensed <sec-yaml-ideal-condensed>`
+- :ref:`ideal-solution-VPSS <sec-yaml-ideal-solution-VPSS>`: An ideal solution;
+  Uses "variable pressure standard state" methods for calculating thermodynamic
+  properties
```
I would like to deprecate this class, but it's more complicated so I think it stays for the time being.

speth commented on pages/science/reactions.rst:
```diff
@@ -13,25 +13,31 @@
+
+.. math::
+
+   \mathrm{A + B \rightleftharpoons C + D}
```
I'm 👎 on changing the names of the generic species. What is here even corresponds to the first time a reaction equation is introduced in _Chemically Reacting Flow_ (eq. 2.27 of the 2nd ed.),  and in that instance it's not seen as necessary to explain what A, B, C, and D are. So I think the basic notation of a chemical equation is something we should be able to assume familiarity with.

bryanwweber replied:
That's fine with me, it would certainly make things more complicated for the other equations.

speth commented on pages/tutorials/yaml/phases.rst:
```diff
@@ -0,0 +1,418 @@
+--------
+
+The following input file defines two equivalent gas phases including all
+reactions and species defined in the input file, (with species and reaction data
```
I think it's important to show that the input file shown does contain ``species`` and ``reactions`` *sections*. It's just that the full data in each section is not shown.

bryanwweber replied:
OK, the purpose of the suggestion was to remove the parenthetical and turn it into a complete sentence. If you remove the parenthetical, the sentence ends with comma-period, which is weird.

bryanwweber commented:
@speth Wow! Was definitely not expecting to change all `e.g.` and `i.e.` that you did. Thank you! I probably should have explained about the shift from `""` to bold - I think the quotes can be interpreted in strange ways, they can come across as scare quotes or something like that (I've learned this from writing for other places). Hence, IMO we should use either bold or italics to denote a **term** that users should be aware of, like a textbook I suppose. Italics are generally used for emphasis, so terms get bolded the first time they're used.

I'll take a look at the rest of this tomorrow. Thanks again!

bryanwweber commented on pages/science/species.rst:
```diff
@@ -107,26 +92,20 @@ Species Transport Coefficients
-``thermo`` field of a :cti:class:`species`. Each implements a different
+The models described in this section can be used to provide thermodynamic data
+for each species in a phase. Each model implements a different
 *parameterization* (functional form) for the heat capacity. Note that there is
```
```suggestion
**parameterization** (functional form) for the heat capacity. Note that there is
```
